metadata/src/main/java/org/apache/kafka/controller/FeatureControlManager.java

```diff
@@ -284,6 +334,120 @@ FinalizedControllerFeatures finalizedFeatures(long epoch) {
+            // Initialize the log with a ZkMigrationState
+            if (zkMigrationEnabled) {
+                log.info("Writing ZkMigrationState of PRE_MIGRATION to the log, since migrations are enabled.");
+                recordConsumer.accept(buildRecord(ZkMigrationState.PRE_MIGRATION));
+            } else {
+                log.debug("Writing ZkMigrationState of NONE to the log, since migrations are not enabled.");
+                recordConsumer.accept(buildRecord(ZkMigrationState.NONE));
+            }
+        } else {
+            // non-empty log
+            String prefix = "During metadata log initialization,";
+            ZkMigrationControlState state = migrationControlState.get();
+            switch (state.zkMigrationState()) {
+                case UNINITIALIZED:
+                    // No ZkMigrationState record has been seen yet
+                    log.debug("{} did not read any ZkMigrationState. Writing a ZkMigrationState of NONE to the log to " +
+                            "indicate this cluster was not migrated from ZK.", prefix);
+                    if (zkMigrationEnabled) {
+                        log.error("Should not have ZK migrations enabled on a cluster that was created in KRaft mode");
+                    }
+                    recordConsumer.accept(buildRecord(ZkMigrationState.NONE));
+                    break;
+                case NONE:
+                    // This is a non-migrated KRaft cluster
+                    log.debug("{} read a ZkMigrationState of NONE indicating this cluster was never migrated from ZK.", prefix);
+                    if (zkMigrationEnabled) {
+                        log.error("Should not have ZK migrations enabled on a cluster that was created in KRaft mode");
+                    }
+                    break;
+                case POST_MIGRATION:
+                    // This is a migrated KRaft cluster
+                    log.debug("{} read a ZkMigrationState of POST_MIGRATION indicating this cluster was previously migrated from ZK.", prefix);
+                    break;
+                case MIGRATION:
+                    // This cluster is migrated, but still in dual-write mode
+                    if (zkMigrationEnabled) {
+                        log.debug("{} read a ZkMigrationState of MIGRATION indicating this cluster is being migrated from ZK.", prefix);
+                    } else {
+                        throw new IllegalStateException(
+                                prefix + " read a ZkMigrationState of MIGRATION indicating this cluster is being migrated " +
+                                        "from ZK, but the controller does not have migrations enabled."
+                        );
+                    }
+                    break;
+                case PRE_MIGRATION:
+                    if (!state.preMigrationSupported()) {
+                        // Upgrade case from 3.4. The controller only wrote PRE_MIGRATION during migrations in that version,
```

Comment:
Hmm, this is a pretty big mess.

I don't think adding a new "preMigrationSupported" field is the right fix, though. Can we just switch the enum values of PRE_MIGRATION and MIGRATION? Since it sounds like 3.4 doesn't use MIGRATION at all, and writes PRE_MIGRATION in all cases.